modeling/algorithm/local-shared-algorithm.go

```diff
@@ -0,0 +1,205 @@
+
+		// calculate expected number of endpoints based on the proportion of
+		// nodes in this zone
+		expectedEndpoints := zone.NodesRatio * float64(region.TotalEndpoints)
+		// deviation: a negative value means need more endpoints from other
+		// sliceGroups, a positive value means need give out endpoints to other
+		// sliceGroups
+		deviation := float64(zone.Endpoints) - expectedEndpoints
+		// if deviation > 0, this zone is a qualified candidate for
+		// availablePool that contributes endpoints to other zones
+		validCandidate := !math.Signbit(deviation)
+		if validCandidate {
+			availablePool.ZoneNames = append(availablePool.ZoneNames, zoneName)
+		}
+		receiverPool.ZoneNames = append(receiverPool.ZoneNames, zoneName)
+		// merge all the zones have no endpoints into a whole
+		if zone.Endpoints == 0 {
+			// this is a form used to accuratly represent the deviation in
+			// float, floatDeviation = 1 * -deviation
+			endpointsNeededUrgent.push(endpointDeviation{name: zoneName, deviation: 1, weight: -deviation})
+			continue
+		}
+		localGroup.Composition[zoneName] = types.WeightedEndpoints{Number: zone.Endpoints, Weight: 1}
+		sliceGroups[zoneName] = localGroup
+
+		// deviation = -0.xx will end up with 0, omit those cases
+		if deviation <= -1 {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: int(-deviation)})
+		}
+	}
+	availablePool.SliceGroups = sliceGroups
+	receiverPool.SliceGroups = sliceGroups
+	heap.Init(&availablePool)
+	heap.Init(&receiverPool)
+
+	succ, err := alg.balanceSliceGroups(&endpointsNeeded, &endpointsNeededUrgent, sliceGroups, &availablePool, &receiverPool)
+	if !succ {
+		klog.Infoln("fail to do local algorithm, switch to shared global algorithm")
+		sharedAlg := SharedGlobalAlgorithm{SharedGlobalAlgorithmCore{globalWeight: 1, globalThreshold: 100}}
+		return sharedAlg.CreateSliceGroups(region)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return sliceGroups, nil
+}
+
+// balanceSliceGroups distributes endpoints from zones with extra endpoints to
+// EndpointSliceGroups for zones with insufficient endpoints.
+func (alg LocalSharedSliceAlgorithm) balanceSliceGroups(endpointsNeeded *endpointsList, endpointsNeededUrgent *endpointsList, sliceGroups map[string]types.EndpointSliceGroup, availablePool *ZonePriorityQueue, receiverPool *ZonePriorityQueue) (bool, error) {
+	// merge one sharedSG that zones in the urgent list will consume
+	mergedSG := types.EndpointSliceGroup{Composition: map[string]types.WeightedEndpoints{}, ZoneTrafficWeights: map[string]float64{}}
+	mergedED := endpointDeviation{name: "merged"}
+	mergedDeviation := 0.0
+	for _, urgentZone := range endpointsNeededUrgent.byZone {
+		mergedED.name += "-" + urgentZone.name
+		mergedDeviation += (float64(urgentZone.deviation) * urgentZone.weight)
+		mergedSG.ZoneTrafficWeights[urgentZone.name] = 1
+		endpointsNeededUrgent.pop()
+	}
+	// take the ceil, in case the deviation = 0.xx, we have to make sure there
```

Comment:
Would this also work?
```suggestion
	// take the ceil, if the deviation > 0, we have to make sure there
```

Reply:
changed